tests/stubtest_third_party.py

```diff
@@ -128,6 +134,81 @@ def run_stubtest(dist: Path, *, verbose: bool = False, specified_platforms_only:
     return True
 
 
+def setup_uwsgi_stubtest_command(dist: Path, venv_dir: Path, stubtest_cmd: list[str]) -> bool:
+    """Perform some black magic in order to run stubtest inside uWSGI.
+
+    We have to write the exit code from stubtest to a surrogate file
+    because uwsgi --pyrun does not exit with the exitcode from the
+    python script. We have a second wrapper script that passed the
+    arguments along to the uWSGI script and retrieves the exit code
+    from the file, so it behaves like running stubtest normally would.
+    """
+    uwsgi_ini = dist / "@tests/uwsgi.ini"
+    if not uwsgi_ini.exists():
+        print_error("Did not find a uwsgi.ini for the uWSGI tests")
+        return False
+
+    uwsgi_script = venv_dir / "uwsgi_stubtest.py"
+    wrapper_script = venv_dir / "uwsgi_wrapper.py"
+    exit_code_surrogate = venv_dir / "exit_code"
+    uwsgi_script_contents = dedent(
+        f"""
+        import json
+        import os
+        import sys
+        from mypy.stubtest import main
+
+        sys.argv = json.loads(os.environ.get("STUBTEST_ARGS"))
+        exit_code = main()
+        with open("{exit_code_surrogate}", mode="w") as fp:
+            fp.write(str(exit_code))
+        sys.exit(exit_code)
+        """
+    )
+    uwsgi_script.write_text(uwsgi_script_contents)
+
+    if sys.platform == "win32":
+        uwsgi_exe = venv_dir / "Scripts" / "uwsgi.exe"
+    else:
+        uwsgi_exe = venv_dir / "bin" / "uwsgi"
+
+    if not uwsgi_exe.exists():
+        print_error("Did not find a uwsgi executable")
+        return False
+
+    # This wrapper script only forwards stdout, since stderr contains
+    # a bunch of spam from uWSGI. If you want to debug the internal script,
+    # it's a good idea to forward stderr temporarily as well
+    wrapper_script_contents = dedent(
+        f"""
+        import json
+        import os
+        import subprocess
+        import sys
+
+        stubtest_env = os.environ | {{"STUBTEST_ARGS": json.dumps(sys.argv)}}
+        uwsgi_cmd = [
+            "{uwsgi_exe}",
+            "--ini",
+            "{uwsgi_ini}",
+            "--spooler",
+            "{venv_dir}",
+            "--pyrun",
+            "{uwsgi_script}",
+        ]
+        ret = subprocess.run(uwsgi_cmd, env=stubtest_env, capture_output=True)
+        print(ret.stdout.decode(), end="", file=sys.stdout)
+        with open("{exit_code_surrogate}", mode="r") as fp:
+            sys.exit(int(fp.read()))
+        """
+    )
+    wrapper_script.write_text(wrapper_script_contents)
```

Comment:
What's the advantage of dynamically creating this script here and then executing it, as opposed to just adding a new script to our `tests/` directory and running that script from a subprocess in this script? :)

Reply:
I suppose there is no real advantage. Since stubtest_third_party was already creating temporary venvs it just seemed natural to make the scripts temporary as well. And the scripts are short enough, that I'm not too worried yet about missing out on the linting / static analysis.

But if you feel strongly enough about it I can certainly be persuaded to restructure this accordingly.

Reply:
Moving it into a static script would also add some minor overhead, if you still wanted to be able to run the script outside the temporary venv created by stubtest_third_party.

Reply:
It struck me that, at this point, we were running stubtest from a script being run from a subprocess triggered from a script dynamically generated by `stubtest_third_party.py`. Turtles all the way down!

But I don't feel too strongly about it :) this is all much more readable now that it's factored out into a separate function! I think I'm basically fine with the structure of your changes as they are now.